ensure errno loaded from msvcrt when checking iconv success

kevinushey opened the pull request:
This should fix #2315.

---

Some background: it looks like `Riconv.dll` imports `msvcrt.dll`, and uses `_errno` from there. On the other hand, `rsession` is getting it from a different runtime altogether: `api-ms-win-crt-runtime-l1-1-0.dll`, which I believe is required for applications built with MSVC 2015 + Qt.

One alternative way of approaching this would be to attempt to link directly with `msvcrt.dll`, but I'm not sure if this would work with MSVC + Qt (although this may be worth exploring if we're not a fan of this solution).

We might want to wrap this up into something cleaner, and make it clear that when calling R APIs that set `errno` we need to explicitly grab it from `msvcrt.dll`.

kevinushey commented:
We should probably audit our usages of `errno` to check whether they're being set by R versus a Windows system call, since we'll need to use this 'hack' to ensure we're reading the right `errno`.

jmcphers commented on src/cpp/r/RUtil.cpp:
```diff
@@ -39,6 +39,20 @@ namespace rstudio {
+
+int errorNumber()
+{
+   return *s_pErrorNumber;
```
This will dereference uninitialized memory if `setErrorNumber` is not called, which isn't likely but looks possible if an error occurs while trying to hook it up. How about init'ing the pointer to `nullptr` and returning a safe error number (0?) when the pointer is not initialized?

kevinushey commented:
Should be good to go now -- I refactored much of the gunk into a separate file / class, and also added `R_ERRNO` in case we need to ever be conscious of the difference between the system-used `errno` and the R-used `errno`.
